src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -835,9 +836,29 @@ private PackageLicenseInfo GeneratePackageLicenseInfo(IPackageSearchMetadata met
             var results = new List<ResolvedAction>();
 
             Debug.Assert(userAction.PackageId != null, "Package id can never be null in a User action");
+
             if (userAction.Action == NuGetProjectActionType.Install)
             {
-                foreach (var target in targets)
+                // find out build integrated projects
+                var buildIntegratedProjectsToUpdate = targets.OfType<BuildIntegratedNuGetProject>().ToList();
+
+                // Other non-builtIntegrated type of projects
+                var otherTargetProjectsToUpdate = targets.Except(buildIntegratedProjectsToUpdate).ToList();
```

Comment:
I don't know how slow this really is, maybe even in a 500 project solution it's insignificant, but this code is iterating `targets` twice, and the second time doing an even more expensive check (O(n^2)).

```suggestion
                var buildIntegratedProjectsToUpdate = new List<BuildIntegratedNuGetProject>();
                var otherTargetProjectsToUpdate = new List<NuGetProject>();
                
                foreach (var proj in targets)
                {
                    if (proj is BuildIntegratedNuGetProject buildIntegratedNuGetProject)
                    {
                        buildIntegratedProjectsToUpdate.Add(buildIntegratedNuGetProject);
                    }
                    else
                    {
                        otherTargetProjectsToUpdate.Add(proj);
                    }
                }
```

Reply:
Taken your suggestion. Thank you.